pkg/controller/elasticsearch/pdb/reconcile.go

```diff
@@ -7,107 +7,165 @@ package pdb
+// Reconcile ensures that a PodDisruptionBudget exists for this cluster, inheriting the spec content.
+// The default PDB we setup dynamically adapts MinAvailable to the number of nodes in the cluster.
+// If the spec has disabled the default PDB, it will ensure none exist.
+func Reconcile(k8sClient k8s.Client, scheme *runtime.Scheme, es v1alpha1.Elasticsearch, statefulSets sset.StatefulSetList) error {
+	expected, err := expectedPDB(es, statefulSets, scheme)
+	if err != nil {
+		return err
 	}
-
-	var objectMeta metav1.ObjectMeta
-	if template != nil {
-		objectMeta = *template.ObjectMeta.DeepCopy()
+	if expected == nil {
+		return deleteDefaultPDB(k8sClient, es)
 	}
-	objectMeta.Name = name.DefaultPodDisruptionBudget(es.Name)
-	objectMeta.Namespace = es.Namespace
-	objectMeta.Labels = defaults.SetDefaultLabels(objectMeta.Labels, label.NewLabels(k8s.ExtractNamespacedName(&es)))
 
-	expected := v1beta1.PodDisruptionBudget{
-		ObjectMeta: objectMeta,
-		Spec:       template.Spec,
+	// reconcile actual vs. expected
+	var actual v1beta1.PodDisruptionBudget
+	err = k8sClient.Get(k8s.ExtractNamespacedName(expected), &actual)
+	if err != nil && apierrors.IsNotFound(err) {
+		return k8sClient.Create(expected)
 	}
 
-	if disabled {
-		// delete the default budget if it exists.
-		//
-		// we do this by getting first because that is a local cache read,
-		// versus a Delete call, which would hit the API.
-
-		if err := c.Get(k8s.ExtractNamespacedName(&expected), &expected); err != nil && !errors.IsNotFound(err) {
-			return err
-		} else if errors.IsNotFound(err) {
-			// already deleted, which is fine
-			return nil
+	// update if we're missing a label
+	needsLabelUpdate := false
+	for k, v := range expected.Labels {
+		if actualValue, ok := actual.Labels[k]; !ok || actualValue != v {
+			needsLabelUpdate = true
 		}
+	}
 
-		if err := c.Delete(&expected); err != nil && !errors.IsNotFound(err) {
+	if needsLabelUpdate || !reflect.DeepEqual(expected.Spec, actual.Spec) {
+		// PDB Spec cannot be updated, we'll have to delete then recreate.
+		// Which means there is a time window in between where we don't have a PDB anymore.
+		// TODO: this is not true anymore starting k8s 1.15+ and this PR https://github.com/kubernetes/kubernetes/pull/69867
+		if err := deleteDefaultPDB(k8sClient, es); err != nil {
 			return err
 		}
+		return k8sClient.Create(expected)
+	}
+
+	return nil
+}
+
+// deleteDefaultPDB deletes the default pdb if it exists.
+func deleteDefaultPDB(k8sClient k8s.Client, es v1alpha1.Elasticsearch) error {
+	// we do this by getting first because that is a local cache read,
+	// versus a Delete call, which would hit the API.
+	pdb := v1beta1.PodDisruptionBudget{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: es.Namespace,
+			Name:      name.DefaultPodDisruptionBudget(es.Name),
+		},
+	}
+	if err := k8sClient.Get(k8s.ExtractNamespacedName(&pdb), &pdb); err != nil && !errors.IsNotFound(err) {
```

Comment:
Is this optimization necessary here?

Reply:
Not strictly necessary, but if we don't do it, and you disable the PDB in the ES spec, we would end up doing a DELETE at every single reconciliation. So I'd say it's worth it.

Reply:
Ah, got it now. Agreed.